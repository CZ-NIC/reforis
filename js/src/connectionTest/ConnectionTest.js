/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, {useState} from 'react';
import ConnectionTestButton from './ConnectionTestButton';
import {useAPIGetData} from '../forisAPI/hooks';
import {useWS} from '../webSockets/hooks';

export const TEST_STATES = {
    NOT_RUNNING: 0,
    RUNNING: 1,
    FINISHED: 2,
};

export default function ConnectionTest() {
    const [state, setState] = useState(TEST_STATES.NOT_RUNNING);
    const initialResults = {
        ipv6: null,
        ipv6_gateway: null,
        ipv4: null,
        ipv4_gateway: null,
    };
    const [testResults, setTestResults] = useState(initialResults);
    const [, setTestID] = useState(null);
    const [getData] = useAPIGetData('connectionTest');

    useWS('wan', 'connection_test', msg => {
        setTestResults(prevTestResults => ({
            ...prevTestResults,
            ...msg.data.data
        }));
    });
    useWS('wan', 'connection_test_finished', msg => {
        setTestResults(msg.data.data);
        setState(TEST_STATES.FINISHED);
    });

    function onSubmit(e) {
        e.preventDefault();
        setState(TEST_STATES.RUNNING);
        setTestResults(initialResults);
        getData(data => {
            setTestID(data.connection_test_id);
        });
    }


    return <form onSubmit={onSubmit}>
        {state !== TEST_STATES.NOT_RUNNING ? <TestResults {...testResults}/> : null}
        <ConnectionTestButton state={state}/>
    </form>
}

const TEST_TYPES = {
    ipv4: _('IPv4 connectivity'),
    ipv4_gateway: _('IPv4 gateway connectivity'),
    ipv6: _('IPv6 connectivity'),
    ipv6_gateway: _('IPv6 gateway connectivity'),
};

function TestResults(tests) {
    return <table className='table table-borderless table-hover offset-lg-3 col-lg-6 col-sm-12'>
        <tbody>
        {Object.keys(tests).map(
            test => {
                const type = TEST_TYPES[test];
                return type ? <TestResultItem key={type} type={type} result={tests[test]}/> : null;
            }
        )}

        </tbody>
    </table>;
}

function TestResultItem({type, result}) {
    let icon = null;
    switch (result) {
        case true:
            icon = <i className='fas fa-times text-danger'/>;
            break;
        case false:
            icon = <i className='fas fa-check text-success'/>;
            break;
        default:
            icon = <div className='spinner-border spinner-border-sm text-secondary' role='status'>
                <span className='sr-only'/>
            </div>;
    }

    return <tr>
        <th scope='row'>{type}</th>
        <td>{icon}</td>
    </tr>
}