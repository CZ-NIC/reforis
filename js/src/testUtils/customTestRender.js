/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

/* eslint import/export: "off" */

import React from 'react';
import {UIDReset} from 'react-uid';
import {StaticRouter} from 'react-router';
import {render} from '@testing-library/react'

function Wrapper({children}) {
    return <StaticRouter>
        <UIDReset>
            {children}
        </UIDReset>
    </StaticRouter>
}

const customTestRender = (ui, options) => render(ui, {wrapper: Wrapper, ...options});

// re-export everything
export * from '@testing-library/react'

// override render method
export {customTestRender as render}
