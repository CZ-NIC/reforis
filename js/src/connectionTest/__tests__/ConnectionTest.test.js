/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import {fireEvent, render, waitForElement} from 'customTestRender';
import mockAxios from 'jest-mock-axios';

import {mockedWS} from 'mockWS';

import ConnectionTest from '../ConnectionTest';

describe('<ConnectionTest/>', () => {
    const mockWebSockets = new mockedWS();

    it('Snapshot before connection test.', () => {
        const {asFragment} = render(<ConnectionTest ws={mockWebSockets} type='wan'/>);
        expect(asFragment()).toMatchSnapshot();
    });

    it('Snapshot after trigger WAN connection test.', async () => {
        const {asFragment, getByText} = render(<ConnectionTest ws={mockWebSockets} type='wan'/>);
        fireEvent.click(getByText('Test connection'));
        mockAxios.mockResponse({data: {test_id: "test-id"}});
        await waitForElement(() => getByText('IPv6 connectivity'));

        expect(mockAxios.post).toBeCalled();
        expect(mockAxios.post).toHaveBeenCalledWith('/api/connection-test', undefined, expect.anything());
        expect(asFragment()).toMatchSnapshot();
    });

    it('Snapshot after trigger DNS connection test.', async () => {
        const {asFragment, getByText} = render(<ConnectionTest ws={mockWebSockets} type='dns'/>);
        fireEvent.click(getByText('Test connection'));
        mockAxios.mockResponse({data: {test_id: "test-id"}});
        await waitForElement(() => getByText(/DNSSEC/));

        expect(mockAxios.post).toBeCalled();
        expect(mockAxios.post).toHaveBeenCalledWith('/api/dns-test', undefined, expect.anything());
        expect(asFragment()).toMatchSnapshot();
    })
});
