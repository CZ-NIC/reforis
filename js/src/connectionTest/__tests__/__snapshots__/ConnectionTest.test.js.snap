// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConnectionTest/> Snapshot after trigger DNS connection test. 1`] = `
<DocumentFragment>
  <form>
    <table
      class="table table-borderless table-hover offset-lg-3 col-lg-6 col-sm-12"
    >
      <tbody>
        <tr>
          <th
            scope="row"
          >
            DNS
          </th>
          <td>
            <div
              class="spinner-border spinner-border-sm text-secondary"
              role="status"
            >
              <span
                class="sr-only"
              />
            </div>
          </td>
        </tr>
        <tr>
          <th
            scope="row"
          >
            DNSSEC
          </th>
          <td>
            <div
              class="spinner-border spinner-border-sm text-secondary"
              role="status"
            >
              <span
                class="sr-only"
              />
            </div>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="col-sm-12 offset-lg-1 col-lg-10 p-0 mb-3 text-right"
    >
      <button
        class="btn btn-primary col-sm-12 col-lg-3"
        disabled=""
        type="submit"
      >
        <span
          aria-hidden="true"
          class="spinner-border spinner-border-sm"
          role="status"
        />
           Test is running...
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`<ConnectionTest/> Snapshot after trigger WAN connection test. 1`] = `
<DocumentFragment>
  <form>
    <table
      class="table table-borderless table-hover offset-lg-3 col-lg-6 col-sm-12"
    >
      <tbody>
        <tr>
          <th
            scope="row"
          >
            IPv4 connectivity
          </th>
          <td>
            <div
              class="spinner-border spinner-border-sm text-secondary"
              role="status"
            >
              <span
                class="sr-only"
              />
            </div>
          </td>
        </tr>
        <tr>
          <th
            scope="row"
          >
            IPv4 gateway connectivity
          </th>
          <td>
            <div
              class="spinner-border spinner-border-sm text-secondary"
              role="status"
            >
              <span
                class="sr-only"
              />
            </div>
          </td>
        </tr>
        <tr>
          <th
            scope="row"
          >
            IPv6 connectivity
          </th>
          <td>
            <div
              class="spinner-border spinner-border-sm text-secondary"
              role="status"
            >
              <span
                class="sr-only"
              />
            </div>
          </td>
        </tr>
        <tr>
          <th
            scope="row"
          >
            IPv6 gateway connectivity
          </th>
          <td>
            <div
              class="spinner-border spinner-border-sm text-secondary"
              role="status"
            >
              <span
                class="sr-only"
              />
            </div>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="col-sm-12 offset-lg-1 col-lg-10 p-0 mb-3 text-right"
    >
      <button
        class="btn btn-primary col-sm-12 col-lg-3"
        disabled=""
        type="submit"
      >
        <span
          aria-hidden="true"
          class="spinner-border spinner-border-sm"
          role="status"
        />
           Test is running...
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`<ConnectionTest/> Snapshot before connection test. 1`] = `
<DocumentFragment>
  <form>
    <table
      class="table table-borderless table-hover offset-lg-3 col-lg-6 col-sm-12"
    >
      <tbody>
        <tr>
          <th
            scope="row"
          >
            IPv4 connectivity
          </th>
          <td>
            <div
              class="spinner-border spinner-border-sm text-secondary"
              role="status"
            >
              <span
                class="sr-only"
              />
            </div>
          </td>
        </tr>
        <tr>
          <th
            scope="row"
          >
            IPv4 gateway connectivity
          </th>
          <td>
            <div
              class="spinner-border spinner-border-sm text-secondary"
              role="status"
            >
              <span
                class="sr-only"
              />
            </div>
          </td>
        </tr>
        <tr>
          <th
            scope="row"
          >
            IPv6 connectivity
          </th>
          <td>
            <div
              class="spinner-border spinner-border-sm text-secondary"
              role="status"
            >
              <span
                class="sr-only"
              />
            </div>
          </td>
        </tr>
        <tr>
          <th
            scope="row"
          >
            IPv6 gateway connectivity
          </th>
          <td>
            <div
              class="spinner-border spinner-border-sm text-secondary"
              role="status"
            >
              <span
                class="sr-only"
              />
            </div>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="col-sm-12 offset-lg-1 col-lg-10 p-0 mb-3 text-right"
    >
      <button
        class="btn btn-primary col-sm-12 col-lg-3"
        type="submit"
      >
          Test connection
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`<ConnectionTest/> Snapshots with connection test results. 1`] = `
<DocumentFragment>
  <form>
    <table
      class="table table-borderless table-hover offset-lg-3 col-lg-6 col-sm-12"
    >
      <tbody>
        <tr>
          <th
            scope="row"
          >
            IPv4 connectivity
          </th>
          <td>
            <div
              class="text-success"
            >
              <i
                class="fas fa-check"
              />
            </div>
          </td>
        </tr>
        <tr>
          <th
            scope="row"
          >
            IPv4 gateway connectivity
          </th>
          <td>
            <div
              class="text-success"
            >
              <i
                class="fas fa-check"
              />
            </div>
          </td>
        </tr>
        <tr>
          <th
            scope="row"
          >
            IPv6 connectivity
          </th>
          <td>
            <div
              class="text-success"
            >
              <i
                class="fas fa-check"
              />
            </div>
          </td>
        </tr>
        <tr>
          <th
            scope="row"
          >
            IPv6 gateway connectivity
          </th>
          <td>
            <div
              class="text-success"
            >
              <i
                class="fas fa-check"
              />
            </div>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="col-sm-12 offset-lg-1 col-lg-10 p-0 mb-3 text-right"
    >
      <button
        class="btn btn-primary col-sm-12 col-lg-3"
        type="submit"
      >
          Test connection again
      </button>
    </div>
  </form>
</DocumentFragment>
`;
